contracts/modules/Experimental/Checkpoints/WeightedVoteCheckpoint.sol

```diff
@@ -0,0 +1,188 @@
+        uint256 totalProposals;
+        bool isActive;
+        mapping(uint256 => uint256) proposalToVote;
+        mapping(address => Vote) voteByAddress;
+    }
+
+    Ballot[] ballots;
+
+    struct Vote {
+        uint256 voteTime;
+        uint256 weight;             
+        uint256 vote;
+    }
+
+    event BallotCreated(uint256 _startTime, uint256 _endTime, uint256 _ballotId, uint256 _checkpointId, uint256 _noOfProposals);
+    event VoteCasted(uint256 indexed _ballotId, uint256 indexed _proposalId, address indexed _investor, uint256 _weight);
+    event BallotStatusChanged(uint256 _ballotId, bool _isActive);
+
+    /**
+     * @notice Constructor
+     * @param _securityToken Address of the security token
+     * @param _polyToken Address of the polytoken
+     */
+    constructor(address _securityToken, address _polyToken)
+    public
+    Module(_securityToken, _polyToken)
+    {
+
+    }
+
+    /**
+     * @notice This function returns the signature of configure function
+     */
+    function getInitFunction() external pure returns(bytes4) {
+        return bytes4(0);
+    }
+
+    /**
+     * @notice Allows the token issuer to create a ballot
+     * @param _duration The duration of the voting period in seconds
+     * @param _noOfProposals Number of proposals
+     */
+    function createBallot(uint256 _duration, uint256 _noOfProposals) external {
+        _onlySecurityTokenOwner();
+        require(_duration > 0, "Incorrect ballot duration.");
+        uint256 checkpointId = ISecurityToken(securityToken).createCheckpoint();
+        uint256 endTime = now.add(_duration);
+        _createCustomBallot(now, endTime, checkpointId, _noOfProposals);
+    }
+
+    function _createCustomBallot(uint256 _startTime, uint256 _endTime, uint256 _checkpointId, uint256 _noOfProposals) internal {
+        require(_noOfProposals > 1, "Incorrect proposals no");
+        require(_endTime > _startTime, "Times are not valid");
+        uint256 ballotId = ballots.length;
+        uint256 supplyAtCheckpoint = ISecurityToken(securityToken).totalSupplyAt(_checkpointId);
+        ballots.push(
+            Ballot(
+                _checkpointId, supplyAtCheckpoint, _startTime, _endTime, 0, _noOfProposals, true
+            )
+        );
+        emit BallotCreated(_startTime, _endTime, ballotId, _checkpointId, _noOfProposals);
+    }
+
+    /**
+     * @notice Allows the token issuer to create a ballot with custom settings
+     * @param _startTime Start time of the voting period in Unix Epoch time
+     * @param _endTime End time of the voting period in Unix Epoch time
+     * @param _checkpointId Index of the checkpoint to use for token balances
+     * @param _noOfProposals Number of proposals
+     */
+    function createCustomBallot(uint256 _startTime, uint256 _endTime, uint256 _checkpointId, uint256 _noOfProposals) external {
+        _onlySecurityTokenOwner();
+        _createCustomBallot(_startTime, _endTime, _checkpointId, _noOfProposals);
```

Comment:
We could explicitly check that `_checkpointId` is valid for a better error message (although `totalSupplyAt` would fail with an invalid checkpoint).